lib/rack/response.rb

```diff
@@ -65,6 +76,8 @@ def chunked?
       CHUNKED == get_header(TRANSFER_ENCODING)
     end
 
+    # @return [Array] a 3-tuple suitable of `[status, headers, body]`
+    # which is a suitable response from the middleware `#call(env)` method.
```

Comment:
This is not true, as I understand, because this method is still returning `self` instead of `body` in the `block_given?` case.

Reply:
And `self` responds to `#each` which is the requirement for `body`.

Reply:
`Response#each` doesn't seem regular…

If you're calling `#finish` with a block, in which violating:

> and must only yield String values

(you can to do so), then SPEC is broken.

I may make mistakes, but… it's hard.

Especially at the end of `body`'s SPEC:

> The Body commonly is an Array of Strings, the application instance itself, or a File-like object.

What of this is a `Response` instance? I know, "commonly", but there is a set…

Reply:
It only needs to respond to `#each` and yield strings. `#finish` with a block yields strings that the user provides. If the user doesn't provide strings, it's fault of user.

Reply:
The correct way to read response body is something like this:

```ruby
status, headers, body = app.call(env)
buffer = String.new
body.each{|chunk| buffer << chunk}
```

I think we should add helper for this in `MockResponse` which has very weird semantics for `#body` and has caused a lot of confusion too.

Reply:
If you're writing something like…

```ruby
result = response.finish { |string| string.upcase! }
result.last.find(&:nil?)
```

You'll get `true` in some (possible unpredictable) cases.

So… it's user's fault to write block (without direct documentation) which returns `nil` sometimes (maybe with one-lined `if`)? I can agree with it, but it looks not enough strict for me.

Reply:
`Rack::Lint` is the place to detect issues like this and it's done this way because it's slower to enforce checks in production in normal code path.